### PR TITLE
Feature/v3

### PR DESCRIPTION
**Pull request by AlbertoFemenias**
## Big changes:
  - Idempotence
  - Tenant structure
  - AUTO_MERGE files
  - No config.yaml

## Usage

There are now 2 versions of the action: `action-state-repo-update-image@feature/v2` and `action-state-repo-update-image@feature/v3`. The purpose is the same but the state repo structure and inputs required differ a lot.

### @feature/v2 - No tenants and use of config.yaml

Version 2 of the action requires to be run in a state repo that follows this structure:
```
state_repo_name
├── app1
│   ├── des
│   │   └── images.yaml
│   ├── pre
│   │   └── images.yaml
│   └── pro
│       └── images.yaml
├── app2
│   ├── dev
│   │   └── images.yaml
│   ├── pre
│   │   └── images.yaml
│   └── pro
│       └── images.yaml
└── config.yaml

```

Note that environment names can be modified as long as they are expressed in the `config.yaml`. The `config.yaml` defines the repo structure and indicates wether image modifications can be automerged or not:

```
# config.yaml
app1:
  auto_merge:
    des: true
    pre: true
    pro: false
  services:
    - app-server
    - app-client
    - app-api

app2:
  auto_merge:
    dev: true
    pre: true
    pro: false
  services:
    - proxy
    - dns

```

You can consume this action by referencing the branch `feature/v2`

```yaml
steps:
  - uses: actions/checkout@v2  #NECESSARY
  - name: Update image  
      uses: prefapp/action-state-repo-update-image@feature/v2
      with:
        application: appfoo
        environment: dev
        services: app-server, app-client # comma separated values, only 1 is needed
        image: image:new
        reviewers: juana, suhermana #OPTIONAL comma separated values, only 1 is needed
        pr_title: PR title #OPTIONAL (not recomended) 
        pr_body: PR body #OPTIONAL (not recomended) 
        branch_name: automated/update-image #OPTIONAL
```

### @feature/v3 - Idempotence, use of tenants, AUTO_MERGE and no config.yaml

In the new version of the action, the applications must belong to a tenant and the automerge is defined
by the inclusion of an `AUTO_MERGE` file in the desired environment. The structure of the repo is all the action needs, the config.yaml is deprecated!

Also, due to the expected external usage, this action is now idempotent. If you try to update an image with the same name and tag it had before nothing will happen: no PR, no commit, just a termination message.

State repo structure required:
```
state_repo_name
├── tenant1
│   ├── release1
│   │   ├── dev
│   │   │   ├── AUTO_MERGE
│   │   │   └── images.yaml
│   │   ├── pre
│   │   │   ├── AUTO_MERGE
│   │   │   └── images.yaml
│   │   └── pro
│   │       └── images.yaml
│   └── release2
│       ├── dev
│       │   ├── AUTO_MERGE
│       │   └── images.yaml
│       ├── pre
│       │   └── images.yaml
│       └── pro
│           └── images.yaml
└── tenant2
    ├── releaseA
    │   ├── dev
    │   │   ├── AUTO_MERGE
    │   │   └── images.yaml
    │   ├── pre
    │   │   ├── AUTO_MERGE
    │   │   └── images.yaml
    │   └── pro
    │       └── images.yaml
    └── releaseB
        ├── dev
        │   ├── AUTO_MERGE
        │   └── images.yaml
        ├── pre
        │   └── images.yaml
        └── pro
            └── images.yaml
```

You can consume this action by referencing the branch `feature/v3`. Careful, **inputs have changed**:
> Remember that inputs in a github action must always be strings.

```yaml
steps:
  - uses: actions/checkout@v2  #NECESSARY
  - name: Update image  
      uses: prefapp/action-state-repo-update-image@feature/v3
      with:
        tenant: tenant_name
        application: app1
        environment: dev
        service_names: '["api", "client"]' # string! example: ${{ toJSON(github.event.client_payload.service_names) }}
        image: image:new
        reviewers: '["juana", "suhermana"]' #OPTIONAL again JSON array
        pr_title: PR title #OPTIONAL (not recomended) 
        pr_body: PR body #OPTIONAL (not recomended) 
        branch_name: automated/update-image #OPTIONAL
```

### CODE REVIEW

**Review by frmadem — COMMENTED** [review]
@AlbertoFemenias we need to pass as a JSON the services right?

**Comment by AlbertoFemenias** [comment]
> @AlbertoFemenias we need to pass as a JSON the services right?

Yes, but the other action is fine as is. Your action returns an object, you can just `${{ toJSON(github.event.client_payload.service_names) }}` and convert to string JSON when calling this action

**Comment by frmadem** [comment]
> > @AlbertoFemenias we need to pass as a JSON the services right?
> 
> Yes @frmadem , but the other action is fine as is. Your action returns an object, you can just `${{ toJSON(github.event.client_payload.service_names) }}` and convert to string JSON when calling this action

OK @AlbertoFemenias LGTM!

File: README.md [hunk 1]
```diff
@@ -4,12 +4,154 @@
...
+```yaml
+steps:
+  - uses: actions/checkout@v2  #NECESSARY
+  - name: Update image  
+      uses: prefapp/action-state-repo-update-image@feature/v3
+      with:
+        tenant: tenant_name
+        application: app1
+        environment: dev
+        service_names: '["api", "client"]' # string! example: ${{ toJSON(github.event.client_payload.service_names) }}
```
Comment (alambike) [inline README.md:143]: service_names now in v3 would be an array of elements

**Review by alambike — CHANGES_REQUESTED** [review]

**Review by frmadem — APPROVED** [review]

**Pull request merged**